nettests/im/whatsapp.go

```diff
@@ -25,7 +28,12 @@ type WhatsAppTestKeys struct {
 }
 
 // GetTestKeys generates a summary for a test run
-func (h WhatsApp) GetTestKeys(tk map[string]interface{}) (interface{}, error) {
+func (h WhatsApp) GetTestKeys(otk interface{}) (interface{}, error) {
+	tk, ok := otk.(map[string]interface{})
+	if !ok {
+		return nil, errors.New("Unexpected test keys format")
```

Comment:
Maybe we want to do this check (which is done for every test) when `GetTestKeys` is called to avoid duplicating it for every test?

Reply:
With the current code, yes. More specifically, with the current code this is a bit of a bummer because all test keys are actually `map[string]interface{}`. Yet, we have this issue:

1. the `map[string]interface{}` is how Go parses a JSON when you don't tell it anything about its structure, hence the typecast here is always gonna succeed;

2. however, for tests that use Go (`ndt7` and `psiphon`, which are already implemented into this engine but not added to this pull request), it's not true that we can cast to `map[string]interface{}` because they're actually Go data structures.

So, my instinct was that we either (1) serialised to JSON also the test keys of other tests and then got `map[string]interface{}` or (2) that the interface of each test needed to be changed and the type cast pushed down to the specific function.

At the same time, yes, this code duplication is annoying. Do you want me to create another method to the API that takes a `map[string]interface{}`, move the cast in generic code, and choose such method or the current one taking `interface{}` depending on the result of the typecast? This would probably tidy up the code a bit, even though it would add one more function to the set of functions in the `Nettest` interface.

Reply:
What if you make the return value of `ndt` and `psiphon` also return `map[string]interface{}` instead of a golang data struct?

Reply:
My immediate instinct is that I don't like this strategy because it would make the implementation of the engine much less obvious to accomodate for an implementation detail of the cli.

I mean, tests written in Go would need to serialise the data structure they return as JSON, then parse such JSON again into a `var v interface{}`, only to store that `v` into `TestKeys` such that `TestKeys` is in the format expected by the cli.

Especially in a scenario where we'll have more and more tests written in Go, this seems to me the sort of complexity that we'll deeply regret adding in the future.

It may be that there is a better way. But I need to spend more time thinking. I believe it may be worth asking the questions whether:

1. we should have in the engine an exported data structure for each experiment written in C++ and parse the results returned by MK into such data structure, then setting `TestKeys` to such data structure, so that later the CLI just needs to perform a type assertion to the test specific structure to get an already organised view of the results (this would also solve some other `XXX` that I have seen in the probe-cli code)

2. find a way for the experiment to return directly the specific data structure rather than making it opaque by setting the `TestKeys` (but this seems to me more of a bummer because we'll lose the generality of running all experiments using generic code)

Do you have any bad feelings towards option 1?

Reply:
(Oh, and @hellais, I think we've hit here the place where the design of experiments written in Go  collides the most with experiments written in C++, so I'm quite glad we've seen this issue now rather than some months in the future, and super good we're discussing it.)

Reply:
So, after thinking a bit more, I guess the right approach is the following: the engine will return a serialised engine as the result of a measurement. The cli will parse such JSON overlying on top of it the keys to which it's interested. This preserves the previous behaviour (simplifies this PR among other things) and is also nice because it reduces the data coupling between the engine and the cli. We don't actually need to have a full understanding of the JSON structure in the cli, rather we just need to understand some fields. So with this design it seems to me we're hitting a sweet spot where everything ends up being simpler.

Reply:
Yes that approach seems reasonable

Reply:
So, returning a serialised JSON is not feasible. Will make the engine API bad and force to do dirty unclean stuff. It's clearly not the right design.

At the same time it does not feel he right design to discard by default the result of a test as a well defined Go data structure. We're currently suffering from bad data schema for more or less all our measurements and we have complexity everywhere because of that. So, we should instead strive to have more well defined data structures. When more tests will be written in Go, we'll have more precise data structures, because Go kinda forces that.

The right solution seems most likely to add glue to the cli that converts the test keys to the format that the CLI wants, forced by the way in which MK works, and, when we'll have all tests written in Go, get rid of such glue and just cast `interface{}` to the proper data structure of each test. An issue should also be created as a reminder.

I'm playing with code doing that.